### DISCUSSION TITLE
Next/Router emit events

### DISCUSSION THREAD

**Discussion by luyizhang** [vercel/next.js#34071] [Help]
I'm writing a custom hook to prevent a user from navigating away if there's unsaved information (in a form) on a page.

To do this, I want to emit a router event when a route change has been cancelled. However, `router.events.emit('routeChangeStart')` causes a a type error TypeError: Cannot read properties of undefined (reading 'shallow')

```
import { useRouter } from 'next/router'
import { useEffect, useCallback } from 'react'

export default function usePreventWindowUnload(preventDefault: boolean) {
  const confirmMessage = 'Are you sure?'
  const router = useRouter()
  
  const onRouterChangeStart = useCallback(() => {
    if (preventDefault) {
      if (window.confirm(confirmMessage)) {
        return true
      }
    }
    // line causing error
    router.events.emit('routeChangeError')
    throw 'cancelled route change'
  ,} [preventDefault])
  
  useEffect(() => {
    router.events.on('routeChangeStart', onRouteChangeStart)
    return () => {
      router.events.off('routeChangeStart', onRouteChangeStart)
    }
  }, [preventDefault])
}
```

How do I fix this?

**Comment by artem-tsys** [comment]
i had the same error.
if you dig a little in the file that generated the error, you can see what exactly caused it.
in my case it was nextjs-progressbar
the plugin cannot complete its work due to an interrupt route

**Comment by a-bugaj** [comment]
if anyone uses `nextjs-progressbar` and has a similar problem to @luyizhang and @artem-tsys , here is an explanation of why this error occurs and how to get around it -> https://github.com/apal21/nextjs-progressbar/issues/70#issuecomment-1052989113

tl;dr

Just use 
```
router.events.emit("routeChangeError",  "your error message", "your-url", { shallow: false });
```

instead of
```
router.events.emit('routeChangeError')
```

**Comment by florianwalther-private** [comment]
Can't you remove the `emit` line because the routing is canceled by the `throw` alone? This is how it works for me.
